lib/srv/forward/sshserver.go

```diff
@@ -851,6 +850,84 @@ func (s *Server) handleConnection(ctx context.Context, chans <-chan ssh.NewChann
 	}
 }
 
+// handleClientChannels handles channel open requests from the remote server.
+func (s *Server) handleClientChannels(ctx context.Context) {
+	forwardedTCPIP := s.remoteClient.HandleChannelOpen(teleport.ChanForwardedTCPIP)
+
+	for {
+		select {
+		case nch := <-forwardedTCPIP:
+			chanCtx, nch := tracessh.ContextFromNewChannel(nch)
+			ctx, span := s.tracerProvider.Tracer("ssh").Start(
+				oteltrace.ContextWithRemoteSpanContext(ctx, oteltrace.SpanContextFromContext(chanCtx)),
+				fmt.Sprintf("ssh.Forward.OpenChannel/%s", nch.ChannelType()),
+				oteltrace.WithSpanKind(oteltrace.SpanKindServer),
+				oteltrace.WithAttributes(
+					semconv.RPCServiceKey.String("ssh.ForwardServer"),
+					semconv.RPCMethodKey.String("OpenChannel"),
+					semconv.RPCSystemKey.String("ssh"),
+				),
+			)
+
+			go func() {
+				defer span.End()
+				s.handleForwardedTCPIPRequest(ctx, nch)
+			}()
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// handleForwardedTCPIPRequest handles remote port forwarding requests.
+func (s *Server) handleForwardedTCPIPRequest(ctx context.Context, nch ssh.NewChannel) {
+	ch, requests, err := nch.Accept()
+	if err != nil {
+		s.log.WithError(err).Errorf("Error accepting %v channel", teleport.ChanForwardedTCPIP)
+		return
+	}
+	go ssh.DiscardRequests(requests)
+	req, err := sshutils.ParseDirectTCPIPReq(nch.ExtraData())
+	if err != nil {
+		s.logErrorToChannel(ch, err, "Unable to parse request.")
+		return
+	}
+
+	// Create context for this channel. This context will be closed when
+	// forwarding is complete.
+	ctx, scx, err := srv.NewServerContext(ctx, s.connectionContext, s, s.identityContext)
+	if err != nil {
+		s.logErrorToChannel(ch, err, "Unable to create connection context.")
+		return
+	}
+	scx.RemoteClient = s.remoteClient
+	scx.ExecType = teleport.ChanDirectTCPIP
+	scx.SrcAddr = sshutils.JoinHostPort(req.Orig, req.OrigPort)
+	scx.DstAddr = sshutils.JoinHostPort(req.Host, req.Port)
+	defer scx.Close()
+
+	ch = scx.TrackActivity(ch)
+
+	// Open a forwarding channel on the client.
+	outCh, outRequests, err := scx.ServerConn.OpenChannel(nch.ChannelType(), nch.ExtraData())
+	if err != nil {
+		s.logErrorToChannel(ch, err, "Error opening %v channel", teleport.ChanForwardedTCPIP)
+		return
+	}
+	go ssh.DiscardRequests(outRequests)
+
+	event := scx.GetPortForwardEvent()
+	if err := s.EmitAuditEvent(ctx, &event); err != nil {
+		s.log.WithError(err).Error("Error emitting audit event")
+		return
+	}
+
+	if err := utils.ProxyConn(ctx, ch, outCh); err != nil {
```

Comment:
Does the SSH protocol prescribe that no data shall be transmitted over stderr in the channel? If so we don't need to bother with forwarding stderr across but we should still discard data from stderrs and/or read from stderr and kill the channel if we get anything.

Reply:
The protocol doesn't specify, but now that I think about it I would be surprised if anything was expected on stderr for a forwarding channel.